### PR TITLE
calculate memory used by parameters, input and gradients

### PR DESCRIPTION
**Pull request by vchuravy**
See the discussion in #2087

### CODE REVIEW

**Comment by piiswrong** [comment]
This may not be accurate since the input parameters can be shared

**Comment by vchuravy** [comment]
@piiswrong This counts every output of a node once, so this should still work for LSTMs? But I haven't looked at one of them to see how their memory structure is.

In general something like this is quite useful, I was just experimenting with a very deep network and the memory usage for gradients and outputs was dominating the memory cost for everything else.

**Comment by piiswrong** [comment]
OK. Please fix the test error and we can merge

File: src/symbol/graph_executor.cc [hunk 1]
```diff
@@ -881,8 +884,9 @@ void GraphExecutor::Print(std::ostream &os) const {
       os << '\n';
     }
   }
-  os << "Total " << (total_allocated_bytes_ >> 20UL) <<" MB allocated\n";
+  os << "Total " << (total_allocated_bytes_ >> 20UL) <<" MiB allocated\n";
   os << "Total " << total_allocated_temp_ <<" TempSpace resource requested\n";
+  os << "Additional " << (memory_used >> 20UL) <<" MiB Memory used\n";
```
Comment (tqchen) [inline src/symbol/graph_executor.cc]: consider put it before the first line, some of the scripts in memory monger relies on the fact that total memory estimation appears in the certain line https://github.com/dmlc/mxnet-memonger/blob/master/memonger.py#L118

**Comment by vchuravy** [comment]
The test failure stems from the fact that `info.data.dtype()` for a gradient `_plus1_0_grad` in the test `test_model_parallel.py`  fluctuates  it is sometimes 0 and on other times it looks like uninitialized/corrupted memory 17431912, 15444848. I wont have time to look into this in detail till next week.

Is https://github.com/dmlc/mxnet/pull/2090/commits/c6065ce74324a77aadced6ff6eb69a07b3b511b0#diff-5947f0cbe18f54861a666a4f683ab378R869 the right way to get the element sizeof?

**Pull request closed without merging**